---
title: "ChIASim: An in silico procedure for generating Chromain interaction data mediated by a protein"
header-includes:
   - \usepackage{setspace}
   - \doublespacing
#output: rmarkdown::html_vignette
output: pdf_document
vignette: >
  %\VignetteIndexEntry{ChIASim}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
fontsize: 12pt  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
```
# Introduction
Many statistical methods and tools have been developed to analyze ChIA-PET, PLAC-seq, or Hi-ChIP data over the past decade. While each new method typically claims superiority than existing ones, accessment of fairness in such studies is not easy without common benchmarking data sets. 

The goal of this package is to provide a mean to integrate benchomarking data sets for fair comarison among studies. In this package, simulated data are generated by imitating biological experimental procedures. The simulation algorithm is designed by mimicking the major steps of the ChIA-PET experiment. Therefore, objective data sets without bias for any of the methods can be generated and used to compare the performance of methods.

Both ChIA-PET and Hi-ChIP are designed to detect chromatin interactions, specifically enhancer-promoter interactions, where a promoter is marked by a transcription starting site (TSS) and a enhancer region is marked by a transcrition factor binding site (TFBS). Although these two technologies differ in efficiency of cell usage (Mumbach et al, 2016), the resulted data contain the same information. Hence, we use ChIA-PET as an example, although the data generated may be considered as an in silico counterpart of either of experimental protocol.
Specifically, our in silico procedure consists of the following five steps.

Step 1: Cross-linking. To imitate the biological experimental step of cross-linking, we first randomly select $N_E$ (set denoted as E) and $N_P$ (set denoted as P) sites from the known sets of TFBS (enhancer) and TSS (promoter) sites, respectively, where the TFBS are determined based on the protein of interest and the TSS sites are obtained from a genome assembly, such as hg19. The reason for this random selection is to reduce the size of the data, but we
maintain the distribution of TFBS/TSS across chromosomes. To induce random collisions (noise, non-promoter-enhancer interactions), we also randomly select $N_\overline{E}$ (set denoted as $\overline{E}$) and $N_\overline{P}$ (set denoted as $\overline{P}$) sites that are neither TFBS nor TSS. Each of the sites in $\{E,\overline{E}\}$ is paired with each of those in $\{P,\overline{P}\}$, completing the cross-linking step. In particular, an $E \times P$ pairing is treated as a "true" promoter-enhancer interacting pair, whereas the other
pairings are simply "noise", representing random collisions.

Step 2: Restriction enzyme digestion. In the biological experimental protocols, the genome may be cut into fragments using a restriction enzyme. In our in silico procedure, we adopt HindIII (AAGCTT) as the default, as it is a frequent choice in real experiments. However, we design ChIA-Sim to be flexible so that users may supply the DNA pattern of any desired enzyme.

Step 3: Ligation. To imitate the step in an biological experiment where ligations between one end of each of two different fragments -- fixed by cross-linking -- take place
to form "loops", ChIA-Sim connects one end of a piece from $\{E,\overline{E}\}$ with one end of a piece from $\{P,\overline{P}\}$, with a preset probability. Users can set these probabilities as necessary, including having different probability for intra-chromosomal pairings (TFBS and TSS sites
from the same chromosome) versus inter-chromosomal pairings (TFBS and TSS sites from two different chromosomes) and different probabilities between the $E \times P$ pairings (true
pairs) versus the other types of pairings (false pairs). These probabilities can be used to control the noise level (i.e. proportion of random collisions), and the degrees of difficulty in
SIC analysis. Parameter values for controlling these probabilities can be set by a user.

Step 4: Sonication. To mimic the sonication (random break) process in ChIA-PET, a Poisson process is applied to each loop formed in the previous step by cutting it into fragments. Specifically, the number of cut sites on a loop is randomly sampled from a Poisson distribution with the mean set to be the length of the loop divided by a typical fragment length -- resulting in a standard sonication -- we set the default value to be 250 bp, but it can be changed by a user. Then the locations of the cut sites are sampled independently from a Uniform distribution. The loop is then "cut" into fragments according to the chosen cut sites, and fragments that contain the ligation sites are kept and sequenced in the next step.

Step 5: Paired-end Sequencing. Reads of a certain length from the two ends of each kept fragments are obtained and mapped to the reference genome. The default on the read length is set in ChIA-Sim, but it can also take a user's input. Anchors (peaks of mapped
fragments) are then identified, and counts of chromatin interactions are recorded. At the end of this step, ChIA-Sim produces a list of interacting pairs with counts, following the format for data from a real ChIA-PET experiment (Table 1 in the main paper).


This package contains the following two functions:  

 - $\textbf{chiaSim}$: This function simulates a set of chromosomal interactions mediated by a specific protein, which can be regarded as the in silico counterparts of ChIA-PET or Hi-ChIP.
 - $\textbf{convertfmt}$: This function allows user to siwtch the output data format between "base" and "BEDPE" without re-running the whole simulation program. The details on the formats are described below.
 
# Installation
ChIASim can be installed from GitHub by devtools::install_git("https://github.com/sl-lin/ChIASim")

ChIASim depends on other packages, i.e., parallel, BSgenome.Hsapiens.UCSC.hg19, and GenomicFeatures. They need to be installed, one may use the commands below, in advance to running ChIASim.  

  if (!requireNamespace("parallel", quietly = TRUE)) install.packages("parallel")
  
  if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
  
  if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)) BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
  
  if (!requireNamespace("GenomicFeatures", quietly = TRUE)) BiocManager::install("GenomicFeatures")
  
 
# ChIASim for data generation



$chiaSim(TFBSfile="POL2", TSSfile=NULL, mean.frag.length = 250, n.cells = 1E5,lp = 0.8, seqlength = 50, N.E=400, N.P=400, N.nE=400, N.nP=400, REp = "AAGCTT", mc.cores.user = 24, beta = 1, gmclass = 1, outputformat= "base", seed=NULL)$

## Input data format and other parameters for chiaSim

$\textbf{TFBSfile}$  is a dataframe containing transcription factor binding sites (TFBS) information for a specific protein of interest. It should contain at least three columns with the first three columns providing information on the chromosome, start site, and end site of each TFBS for this protein (see example below). The ChIA-Sim package has two built-in TFBS data sets available, "ERa" and "POL2". If interested in a different protein, a user-supplied file may be used as long as it contains the required information.

Here is the built-in TFBS file for POL2:  
```{r}
library(ChIASim)
data("POL2")
head(POL2)
```


$\textbf{TSSfile}$ is a dataframe containing transcription start sites (TSS) information. Users do not need to provide a TSS file as the package has the TSS information of hg19 from Refseq (Pruitt et al, 2014). However, if users do provide a TSS file, it should follow the format as show in the following example.  

```{r}
data("TSS")
head(TSS)
```
In this example, the first column is the gene ID for each TSS. The second column denotes the chromosome in which the gene is located. The third column is strand orientation: + denote forward and - denote reverse. The forth column and the fifth column are the TSS for the forward and reverse strands, respectively. So if a read is from a forward strand, then the TSS is the number in the forth column; otherwise, the TSS is the number in the fifth column. The sixth column is the gene name, which is optional.

The following describes other input parameters.
$\textbf{mean.frag.length}$ is the mean chromosome fragment length for sonication simulation. To simulate the sonication step in a real ChIA-PET experiment that cuts choromosomes into pieces, we need to specify the average length of fragments. The default is set to be 250bp.  

$\textbf{n.cells}$ is the simulated number of cells which may be interpreted as the sequencing depth. By adjusting n.cells, different numbers of pairs can result. The default is set to be 10^5.

$\textbf{lp}$ is the ligation probability. This is the probability that two selected segments to ligate. It is set to 0.8 by default.

$\textbf{seqlength}$ is the length of the fragments to be seqenced. ChIP-PET seqences the two end segments of an interacting pair.  The typical length of PET seqencing is 50-100bp (Fullwood et al, 2009), hence it is set to be 50bp as the default seqencing length.  

$\textbf{N.E}$ ($N_E$ in main paper) is the number of TFBS sites randomly selected from the provided TFBS file for forming true pairs for simulation. The selected TFBS sites are allocated to
each choromosome proportional to their lengths. It is set to 400 by default.
$\textbf{N.P}$ ($N_P$ in the main paper) is the number of TSS sites randomly selected from the provided TSS file for forming true pairs for simulation. The selected TSS sites are allocated to each choromosome proportional to their lengths. It is set to 400 by default.
$\textbf{N.nE}$ ($N_{\overline{E}}$ in the main paper) is the number of non-TFBS sites randomly selected across the whole genome for forming the false pairs for simulation. The selected non-TFBS sites are allocated to each chromosome proportional to their lengths. It is set to 400 by default.
$\textbf{N.nP}$ ($N_{\overline{P}}$ in main paper) is the number of non-TSS sites randomly selected across the whole genome for forming the false pairs for simulation. The selected non-TSS sites are allocated to each choromosome proportional to their lengths. It is set to 400 by default.

$\textbf{REp}$ is the pattern of restriction enzyme. It is set to AAGCTT by default, which is the pattern of HindIII. 
  
$\textbf{mc.cores.user}$ is the number of cores used to do parallel computing when multi cores are available at the user's end.  

$\textbf{beta}$ is the power parameter. Power parameter is the power based on the widely believed power law between interaction frequency and 1/log(genomic distance). It is usually set to 1 for human (Fudenberg and Mirny, 2012), and is the default value.

$\textbf{gmclass}$ is a flagging variable on whether to identify self-ligation pairs. When it is set to 2, all resulting interaction pairs will be clustered into 2 groups according to their genomic distances; the cluster with smaller genomic distance will be treated as self-ligation and removed. When self-ligation is not a major issue, it is set to 1, the default value. 

$\textbf{outputformat}$ is the format to output the simulated data. Users can choose either "base" or "BEDPE". We provide these two formats becasue they are the two most popular data formats for real data. For example, the K562 CTCF ChIA-PET data set with accession number GSM970216 uses the BEDPE format, while the K562 POL2 ChIA-PET data set with accession number GSM832465 uses a 7-column base format. Defualt is set to the base format. We provide two examples in the following, one with the base format and the other with the BEDPE format.  

$\textbf{seed}$ is the random seed number. It may be set for reproducibility purpose. Otherwise, it should be set as null. 

## Example
```{r, message=FALSE}
#The parameters are set to be smaller than default for a quick illustration.
#base format
output.base<- chiaSim(n.cells = 500, N.E=100, N.P=100, N.nE=100, N.nP=100,  mc.cores.user = 1)
head(output.base)
```
The first seven columns are the 7-column base format, which provide the location information for two anchors of an interacting pair plus the interaction count. The additional columns are useful for some analytical tools. Specifically, the LT column is the ligation type, which tells the category of each pair, TL (true high), TL  (true low) and F (false). The MC column is the margincal counts which is the sum of counts of two anchors of a pair interacting with all the other anchors in the data set. Mdist tells the minimum distance of each pair to their closes pair of TFBS and TSS. Gdist is the genomic distance between the midpoints of two interacting anchors. See Lou (2021) for more details.

```{r, message=FALSE}
#BEDPE format
output.BEDPE<- chiaSim(n.cells = 500, N.E=100, N.P=100, N.nE=100, N.nP=100, mc.cores.user = 1, outputformat="BEDPE")
head(output.BEDPE[[1]])
head(output.BEDPE[[2]])
```
The first element in the output list, $output.BEDPE[[1]]$, show the BEDPE data format, which provide the location information for the two anchors of an interacting pair plus the interaction count.  The second element in the output list, $output.BEDPE[[2]]$, contains the information provided from the simulation procedure, i.e., LT, MC, Mdist and Gdist, as described above.  

# Conversion between the two data formats
$\textbf{convertfmt}$: This function allows user to convert the output data from one format to the onther format without having to re-run the whole simulation program.

$convertfmt(input, inputformat)$

$\textbf{input}$ is the dataset which needs to be converted to a different format.

$\textbf{inputformat}$ is the format of the data set to be converted. It can be either "base" or "BEDPE". If it is "base", it will be conveted to BEDPE, and vice versa.

```{r, message=FALSE}
#Converting base to BEDPE
output<- convertfmt(input= output.base, inputformat="base")
head(output[[1]])
head(output[[2]])
```

```{r, message=FALSE}
#Converting BEDPE to base
output<- convertfmt(input= output.BEDPE, inputformat="BEDPE")
head(output)
```

# References
Lou, S. Bayesian Analysis for Significant Interactions of Chromatins
and Simulation Algorithm. Doctoral Dissertation, (2021).

Mumbach, M., Rubin, A., Flynn, R. et al. HiChIP: efficient and sensitive analysis of protein-directed genome architecture. Nat Methods 13, 919-922 (2016). 

Fudenberg, Geoffrey, and Leonid A Mirny. "Higher-order chromatin structure: bridging physics and biology." Current opinion in genetics & development vol. 22,2 (2012).

Pruitt, K. D., G. R. Brown, S. M. Hiatt, F. Thibaud-Nissen, A. Astashyn, O. Ermolaeva, C. M. Farrell, J. Hart, M. J. Landrum, K. M. McGarvey, M. R. Murphy, N. A. O'Leary, S. Pujar, B. Rajput, S. H. Rangwala, L. D. Riddick, A. Shkeda, H. Sun, P. Tamez, R. E. Tully, C. Wallin, D. Webb, J. Weber, W. Wu, M. DiCuccio, P. Kitts, D. R. Maglott, T. D. Murphy and J. M. Ostell (2014): "Refseq: an update on mammalian reference sequences," Nucleic Acids Res., 42, D756-D763. 

Fullwood MJ, Wei CL, Liu ET, Ruan Y. Next-generation DNA sequencing of paired-end tags (PET) for transcriptome and genome analyses. Genome Res. 2009;19(4):521-532. doi:10.1101/gr.074906.107


